server/api/totp/enroll/service.js

```diff
@@ -0,0 +1,83 @@
+import { ObjectId } from 'mongodb';
+import QRCode from 'qrcode';
+import { addSeconds, isBefore } from 'date-fns';
+import {
+  DuplicateAuthFactor,
+  UserNotFoundError,
+  UserDeactivatedError,
+} from '../../../constants/errors';
+
+const TOKEN_LIFETIME_IN_SECONDS = 15 * 60; // 15 mins
+
+export const enrollTOTPAuthFactor = async ({ db, config }, { userId }) => {
+  const UserModel = db.model('User');
+  const TokenModel = db.model('Token');
+  const TOTPModel = db.model('TOTP');
+
+  // retrieve user from db
+  const user = await UserModel.findOneWithAuthFactors({ _id: ObjectId(userId) });
+
+  // make sure user exists
+  if (!user) {
+    throw new UserNotFoundError(`User ${userId} not found`);
+  }
+
+  // make sure user is active
+  if (!!user.deactivatedAt && isBefore(user.deactivatedAt, new Date())) {
+    throw new UserDeactivatedError(`User ${userId} is deactivated`);
+  }
+
+  // ensure user has not already activated TOTP authentication factor
+  if (user.authFactors.some((e) => e.type === 'TOTP')) {
+    throw new DuplicateAuthFactor(`User ${userId} is already enrolled to TOTP authentication`);
+  }
+
+  // generate secret
+  const secret = TOTPModel.generateSecret();
+
+  // update db
+  const session = await db.startSession();
+  session.startTransaction();
+
+  try {
+    // delete previous in-flight totp_enroll tokens (if any)
+    await TokenModel.deleteMany({
+      type: 'TOTP_ENROLL',
```

Comment:
can we maybe eject these strings into a constant variable that we import wherever we need? Its fine now but i don't very much like magic

Reply:
Not magic, just fine-grained business-related logic. I am not sure how to make this clearer.

Reply:
Fixed.